File: python/1_two_sum/main.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        indeces = []
        for i in range(len(nums)):
            diff = target - nums[i]
            nnums = nums[i + 1 :]
            if diff in nnums:
                index = nums.index(diff, i + 1)
                indeces.append(i)
                indeces.append(index)
                break
        return indeces

    def twoSumLR(self, nums: list[int], target: int):
        nums.sort()
        lptr = 0
        rptr = -1
        for _ in range(len(nums)):
            print(nums[lptr] + nums[rptr])
            if nums[lptr] + nums[rptr] < target:
                lptr += 1
            elif nums[lptr] + nums[rptr] > target:
                rptr -= 1
            else:
                print(lptr, rptr)
                return [lptr, len(nums) + rptr]
        return []

    def twoSumBest(self, nums: list[int], target: int):
        seen = {}
        for i, num in enumerate(nums):
            diff = target - num
            if diff in seen:
                return [i, seen[diff]]
            else:
                seen[num] = i
        return []


if __name__ == "__main__":
    sol = Solution()
    print(sol.twoSumBest([3, 2, 4, 9], 6))
    print(sol.twoSumBest([3, 3], 6))
    print(sol.twoSumBest([2, 5, 5, 11], 7))

File: python/1_two_sum/main.c
/*
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
*/

#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct newNode {
  int key;
  int value;
  struct newNode *next;
} newNode;

static uint64_t hash_key(const int key) {
  const uint64_t FNV_OFFSET = 14695981039346656037UL;
  const uint64_t FNV_PRIME = 1099511628211UL;
  uint64_t hash = FNV_OFFSET;
  hash *= FNV_PRIME;
  hash ^= (uint64_t)(key);
  return hash;
}

static void destroy_set(newNode **set, size_t set_size) {
  for (int i = 0; i < set_size; i++) {
    newNode *current_ll = set[i];
    while (current_ll != NULL) {
      newNode *temp = current_ll;
      current_ll = current_ll->next;
      free(temp);
    }
  }
  free(set);
}

int key_in_hashmap(int val, newNode **arr, size_t arr_size) {
  for (int i = 0; i < sizeof(arr_size); i++) {
    uint64_t hash = hash_key(val);
    int index = (hash % arr_size);
    newNode *current = arr[index];
    while (current != NULL) {
      if (current->key == val) {
        return current->value;
      }
      current = current->next;
    }
  }
  return -1;
}

int *twoSum(int *nums, int numsSize, int target, int *returnSize) {
  newNode **linkedLists = calloc(numsSize, sizeof(newNode *));
  int *ret = malloc(*returnSize * sizeof(int));
  for (int i = 0; i < numsSize; i++) {
    int diff = target - nums[i];
    int found = key_in_hashmap(diff, linkedLists, numsSize);
    if (found != -1) {
      ret[0] = i;
      ret[1] = found;
      return ret;
    } else {
      uint64_t hash = hash_key(nums[i]);
      int index = (hash % numsSize);
      newNode *current = linkedLists[index];
      while (current != NULL) {
        current = current->next;
      }
      newNode *node = malloc(sizeof(newNode));
      node->key = nums[i];
      node->value = i;
      node->next = linkedLists[index];
      linkedLists[index] = node;
    }
  }
  return ret;
}

int main(void) {
  int test[] = {2, 7, 11, 15};
  int size = 2;
  int* ret = twoSum(test, sizeof(test) / sizeof(int), 9, &size);
  printf("[%d, %d]\n", ret[0], ret[1]);

}
